// kyber-fpr-sdk/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
        value: true
});
exports.MultiHttpProvider = exports.Reserve = exports.SanityRatesContract = exports.ReserveContract = exports.CompactDataLocation = exports.RateSetting = exports.StepFunctionDataPoint = exports.TokenControlInfo = exports.CompactData = exports.ConversionRatesContract = exports.BaseContract = exports.Deployer = undefined;

var _deployer = require('./deployer.js');

Object.defineProperty(exports, 'Deployer', {
        enumerable: true,
        get: function get() {
                return _interopRequireDefault(_deployer).default;
        }
});

var _base_contract = require('./base_contract.js');

Object.defineProperty(exports, 'BaseContract', {
        enumerable: true,
        get: function get() {
                return _interopRequireDefault(_base_contract).default;
        }
});

var _conversion_rates_contract = require('./conversion_rates_contract.js');

Object.defineProperty(exports, 'ConversionRatesContract', {
        enumerable: true,
        get: function get() {
                return _interopRequireDefault(_conversion_rates_contract).default;
        }
});
Object.defineProperty(exports, 'CompactData', {
        enumerable: true,
        get: function get() {
                return _conversion_rates_contract.CompactData;
        }
});
Object.defineProperty(exports, 'TokenControlInfo', {
        enumerable: true,
        get: function get() {
                return _conversion_rates_contract.TokenControlInfo;
        }
});
Object.defineProperty(exports, 'StepFunctionDataPoint', {
        enumerable: true,
        get: function get() {
                return _conversion_rates_contract.StepFunctionDataPoint;
        }
});
Object.defineProperty(exports, 'RateSetting', {
        enumerable: true,
        get: function get() {
                return _conversion_rates_contract.RateSetting;
        }
});
Object.defineProperty(exports, 'CompactDataLocation', {
        enumerable: true,
        get: function get() {
                return _conversion_rates_contract.CompactDataLocation;
        }
});

var _reserve_contract = require('./reserve_contract.js');

Object.defineProperty(exports, 'ReserveContract', {
        enumerable: true,
        get: function get() {
                return _interopRequireDefault(_reserve_contract).default;
        }
});

var _sanity_rates_contract = require('./sanity_rates_contract.js');

Object.defineProperty(exports, 'SanityRatesContract', {
        enumerable: true,
        get: function get() {
                return _interopRequireDefault(_sanity_rates_contract).default;
        }
});

var _reserve = require('./reserve.js');

Object.defineProperty(exports, 'Reserve', {
        enumerable: true,
        get: function get() {
                return _interopRequireDefault(_reserve).default;
        }
});

var _multiple_http_provider = require('./multiple_http_provider.js');

Object.defineProperty(exports, 'MultiHttpProvider', {
        enumerable: true,
        get: function get() {
                return _interopRequireDefault(_multiple_http_provider).default;
        }
});

require('babel-polyfill');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
